### PR TITLE
Avoid unnecessary sidecarscope calculation

### PR DESCRIPTION
**Pull request by LynnnnnnnYang**
This PR avoid recalculation of unrelated sidecar scopes which are not affected by the destinationrule or virtualservice updated.
It will significantly reduce the time it takes to do InitContext, especially when facing large scale of services and sidecars scenario. 
In our case (1k sidecars and 100k services), it used to take almost 3m to simply do initSidecarScopes. Now it only takes seconds to do initContext, which is mainly depends on how many sidecarscopes are affected.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[X] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by LynnnnnnnYang** [comment]
> @googlebot I signed it!

@googlebot I signed it!

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -1361,6 +1366,40 @@ func (ps *PushContext) initDefaultExportMaps() {
...
+			continue
+		}
+		for _, hostname := range strings.Split(conf.Hostnames, ",") {
+			relatedHostnames[hostname] = struct{}{}
+		}
+	}
+
+	isAffected := func(sc *SidecarScope) bool {
+		for hostname := range relatedHostnames {
+			if sc.DependsOnHostname(hostname) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:1384]: The  hostname  is not fqdn, so this is not quite right

File: pilot/pkg/model/config.go [hunk 2]
```diff
@@ -51,6 +54,44 @@ func (key ConfigKey) HashCode() uint32 {
...
+	key := ConfigKey{
+		Kind:      curr.GroupVersionKind,
+		Name:      curr.Name,
+		Namespace: curr.Namespace,
+	}
+	hostnames := map[string]struct{}{}
+	switch curr.GroupVersionKind {
+	case gvk.VirtualService:
+		if vs, ok := curr.Spec.(*networking.VirtualService); ok {
+			for _, h := range vs.Hosts {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/config.go:69]: Yeah, this is only part of the hosts

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:1384]: never mind, we have resolved hosts to FQDN in initVirtualServices

**Comment by LynnnnnnnYang** [comment]
@googlebot I signed it!

**Pull request closed without merging**